### PR TITLE
T/61: Handling with duplicate packages.

### PR DESCRIPTION
**Pull request by pomek**
Fixes: #61.

### CODE REVIEW

File: lib/tasks.js [hunk 1]
```diff
@@ -159,11 +165,18 @@ const tasks = {
 		 * @param {Object.<String, String>} options.formats Object containing information about output path for each supported format.
 		 * @param {Array.<String>} options.packages Paths to CKEditor 5 dependencies.
 		 * @param {Boolean} options.watch Whether to watch the files.
+		 * @param {Boolean} options.ignoreDuplicates Whether to ignore duplicate packages.
 		 * @returns {Stream}
 		 */
 		sass( options ) {
+			if ( !options.ignoreDuplicates && utils.hasDuplicatedPackages( options.packages ) ) {
```
Comment (Reinmar) [inline lib/tasks.js]: This and the following 4 lines should be wrapped into a single function. DRY.

Comment (Reinmar) [inline lib/tasks.js]: Let's call this function `validatePackages( opts )`. It should return an array of validated (deduped) packages.

File: lib/tasks.js [hunk 2]
```diff
@@ -67,6 +67,7 @@ const tasks = {
 		 * @param {Boolean} options.watch Whether to watch the files.
 		 * @param {Boolean} options.es5 Whether to compile the code to ES5.
 		 * @param {String} options.samplesGlob Glob describing sample files to process.
+		 * @param {Boolean} options.ignoreDuplicates Whether to ignore duplicate packages.
```
Comment (Reinmar) [inline lib/tasks.js]: "duplicated". And add another sentence – "Packages can be duplicated if there are conflicts in dependency versions or when one of the packages is installed in the development mode (as a cloned repository)."

File: README.md [hunk 3]
```diff
@@ -90,6 +90,9 @@ Example configuration:
 
 If you also plan to use the `--es5` option of the compiler, install `babel-polyfill`. It is needed in the project which contains Bender.js configuration, so Bender.js can find it.
 
+By default this package throws an error when given packages are duplicated. You can ignore this error using the option `--ignore-duplicates`.
```
Comment (Reinmar) [inline README.md]: This package does not throw. The compiler does. Besides "when there are duplicates in the given package list".

File: lib/tasks.js [hunk 4]
```diff
@@ -382,3 +391,17 @@ function getSourcesStream( packages, watch, samplesGlob = null ) {
 
 	return merge.apply( null, streams );
 }
+
+/**
```
Comment (Reinmar) [inline lib/tasks.js]: You documented it as a public function when it's not. And in fact it should be public... I don't know why the less useful functions are, when this isn't.

File: tests/tasks.js [hunk 5]
```diff
@@ -219,6 +233,7 @@ body {
 		// See: #48
 		it( 'should return empty stream when entry points were not found', ( done ) => {
 			sandbox.stub( gutil, 'log' );
+			sandbox.stub( utils, 'hasDuplicatedPackages' ).returns( false );
```
Comment (Reinmar) [inline tests/tasks.js]: You should be stubing `validatePackages()`.

Comment (Reinmar) [inline tests/tasks.js]: Or rather – I'd write tests with a valid input so you don't need to stub too many things.

File: lib/utils.js [hunk 6]
```diff
@@ -697,7 +702,86 @@ require( [ 'tests' ], bender.defer(), function( err ) {
...
 };
 
 module.exports = utils;
+
+// Tries to find package name from given path.
+//
+// @param {String} packagePath
+// @returns {String|null}
+function findPackageName( packagePath ) {
+	const matches = packagePath.match( /ckeditor5-([A-Z0-9-_]+)/i );
```
Comment (Reinmar) [inline lib/utils.js]: Actually, this regexp is bad cause it will fail on Windows paths. So it's better to do `path.split()` or something.

Comment (Reinmar) [inline lib/utils.js]: When fixing this, please make sure that the code won't fail on `/foo/ckeditor5-x/`.

File: lib/tasks.js [hunk 7]
```diff
@@ -159,11 +162,14 @@ const tasks = {
 		 * @param {Object.<String, String>} options.formats Object containing information about output path for each supported format.
 		 * @param {Array.<String>} options.packages Paths to CKEditor 5 dependencies.
 		 * @param {Boolean} options.watch Whether to watch the files.
+		 * @param {Boolean} options.ignoreDuplicates Whether to ignore duplicated packages.
 		 * @returns {Stream}
 		 */
 		sass( options ) {
+			const packages = validatePackages( options );
```
Comment (pomek) [inline lib/tasks.js]: https://github.com/ckeditor/ckeditor5-dev-compiler/pull/62#discussion_r86965330 … ?

File: lib/tasks.js (hunk 4 above)
Comment (pomek) [inline lib/tasks.js]: Give me a link to documentation where it is described. I've never read things you mentioned.

File: lib/utils.js (hunk 6 above)
Comment (pomek) [inline lib/utils.js]: Correct regexp is `\/ckeditor5-([^/]+)\/?$`, not `/\/ckeditor5-([^/]+)/?$/`. I will split path using system separator (`path.sep`).

File: lib/tasks.js (hunk 7 above)
Comment (Reinmar) [inline lib/tasks.js]: Sorry, my bad. I wasn't then really considering what's needed. But the packages make sense to be passed alone and then the other options.

File: lib/tasks.js (hunk 4 above)
Comment (Reinmar) [inline lib/tasks.js]: There's no documention about writing documentation yet, cause we don't have it standardised yet (it's standardised for CKE4). But we were never documenting things which are totally private (are not exposed in any way) using `/** */` comments. If anything, we use `//`.

**Comment by pomek** [comment]
I introduced fixes to your notices. Ready to review again.

**Comment by Reinmar** [comment]
I wanted to refactor the code "a bit" and ended changing a lof of it. Please check those changes.

**Pull request merged**